src/clojure/nrepl/cmdline.clj

```diff
@@ -239,64 +249,190 @@
-                                       {:transport transport}))
-            ;; need to hold process open with a non-daemon thread -- this should end up being super-temporary
-            (Thread/sleep Long/MAX_VALUE)))))))
+        merge-config (partial merge config/config)
+        options (-> options
+                    (keywordize-options)
+                    (parse-cli-values)
+                    (merge-config))]
+    [options _args]))
+
+(defn display-help
+  "Displays help copy to the user and exits the program"
+  []
+  (println (help))
+  (exit 0))
+
+(defn display-version
+  "Displays the version to the user and exits the program"
+  []
+  (println (:version-string version/version))
+  (exit 0))
+
+(defn- resolve-transport
+  "Takes a map of nREPL CLI options.
+  Returns either a default transport or the
+  value of :transport."
+  [options]
+  (or (some->> options
+               (:transport)
+               (require-and-resolve :transport))
+      #'transport/bencode))
+
+(defn- resolve-handler
+  "Takes a map of nREPL CLI options and list of middleware.
+  Returns a request handler function.
+  If some handler was explicitly passed we'll use it, otherwise we'll build
+  one from whatever was passed via --middleware"
+  [options middleware]
+  (or (some->> options
+               (:handler)
+               (require-and-resolve :handler))
+      (build-handler middleware)))
+
+(defn- resolve-ack-port
+  "Takes a map of nREPL CLI options.
+  Returns integer ack port or nil."
+  [options]
+  (some-> options
+          (:ack)
+          (->int)))
+
+(defn- resolve-greeting
+  "Takes a map of nREPL CLI options and the selected transport for the server.
+  Returns a greeting function or nil."
+  [options transport]
+  (when (= transport #'transport/tty)
+    #'transport/tty-greeting))
+
+(defn connection-opts
+  "Takes map of nREPL CLI options
+  Returns map of processed options used to connect or start a nREPL server."
+  [options]
+  {:port (->int (:port options))
+   :host (:host options)
+   :transport (resolve-transport options)})
+
+(defn server-opts
+  "Takes a map of nREPL CLI options
+  Returns map of processed options to start an nREPL server."
+  [options]
+  (let [middleware (sanitize-middleware-option (:middleware options))
+        {:keys [host port transport]} (connection-opts options)]
+    (merge options
+           {:host host
+            :port port
+            :transport transport
+            :bind (:bind options)
+            :middleware middleware
+            :handler (resolve-handler options middleware)
+            :greeting (resolve-greeting options transport)
+            :ack-port (resolve-ack-port options)})))
+
+(defn connect-to-server
+  "Connects to a running nREPL server and runs a REPL. Exits program when REPL
+  is closed.
+  Takes a map of nREPL CLI options."
+  [options]
+  (let [{:keys [host port transport]} (connection-opts options)]
```

Comment:
Probably you can move the destructuring to the signature and just move `connection-opts` to the call site of this function.